### PR TITLE
Part Deux!

### PR DESCRIPTION
**Pull request by CCSt130**
Inadvertently committed pt2 while it was in the wrong directory. Sorry for any inconvenience.

### CODE REVIEW

**Comment by natasha-aleksandrova** [comment]
@CCSt130 is this supposed to be lesson 4?

**Comment by CCSt130** [comment]
Yes, sorry for the inconvenience; I forgot to create a new folder.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

File: students/CCSt130/lesson03/mail_room_pt2.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+"""
+Lesson04 :: Mailroom Part 2
+@author: Chuck Stevens :: CCSt130
+Edited Sun Aug 4 17:31:12 2019
+
+"""
+
+import datetime # dates needed for comparison
+
+from collections import Counter # Count of donations
```
Comment (charlesrobison) [inline students/CCSt130/lesson03/mail_room_pt2.py]: Make sure imports are at top of file, with no blank lines between

File: students/CCSt130/lesson03/mail_room_pt2.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+        # Error handling
+        if not usr_input in menu_dict:
+            print()                
+            print("Invalid entry--try again!")
+            print()
+        # Launch a valid option
+        # This will unintentionally print primary keys w/o elif '4'
+        else: menu_dict.get(usr_input)(all_donations)
+
+def quit_menu(exit_strings):
```
Comment (charlesrobison) [inline students/CCSt130/lesson03/mail_room_pt2.py]: You'll want 2 blank lines between functions

File: students/CCSt130/lesson03/mail_room_pt2.py [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+   
+    # List all relevant donations in the email generated
+    list_donor_donations(all_donations, a_donor)
+    
+    # Print signature block   
+    print(donor_email['Sig_block'])
+    
+    return
+
+def prn_ltrs(all_donations):
```
Comment (charlesrobison) [inline students/CCSt130/lesson03/mail_room_pt2.py]: Ideally, your function names and variables should be self explanatory and explicit.  i.e., print_letters()

File: students/CCSt130/lesson03/mail_room_pt2.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+    """ Creates formatted letters with donation details. """
+    # For today's date to be printed in email
+    from datetime import date
```
Comment (charlesrobison) [inline students/CCSt130/lesson03/mail_room_pt2.py]: Go ahead and add this to the imports at the top of the file

File: students/CCSt130/lesson03/mail_room_pt2.py [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+            print("Donor letter has been saved to current directory for: {}!".format(a_donor))
+
+    return
+
+def prn_dict(a_dict):
+    """ Generic dictionary printer """
+    for k, v in a_dict.items():
+        print(k,":",v)
+
+def findKey(ttl):
```
Comment (charlesrobison) [inline students/CCSt130/lesson03/mail_room_pt2.py]: This should be snake_case, just like the rest of your functions and variables

File: students/CCSt130/lesson03/mail_room_pt2.py [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+            print()
+            print("Invalid entry--try again!") 
+            print()
+    # Inform user
+    print()
+    print(">>> Donation Has Been Entered for {}. <<<".format(a_donor))
+    print()
+    print("###")
+    
+    return(a_donor) # Return user's choice
```
Comment (charlesrobison) [inline students/CCSt130/lesson03/mail_room_pt2.py]: This function breaks when adding new donors and their donations, do you know why?

**Comment by CCSt130** [comment]
Yes, that’s correct. I accidentally committed with the file in the wrong directory. Sorry for the confusion. I will fix the donor add issue and resubmit. Thanks, Chuck

Sent from my iPhone

On Aug 7, 2019, at 10:14 PM, Charles Robison <notifications@github.com<mailto:notifications@github.com>> wrote:


@charlesrobison approved this pull request.

@CCSt130<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCCSt130&data=02%7C01%7C%7C921f6ca520154e5b6ef008d71bbf43a4%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637008380595016506&sdata=KBzHR%2BDFhTEdgOwlXG8atJQHUoZY%2FVSUDnSEbi0GUXs%3D&reserved=0> I believe both of these files are the same, for Lesson 4, correct? Just making sure. There are a few PEP8 formatting issues here, but you should be able to spot them quickly with a good linter. The assignment works quite well, but breaks when adding new donors. Please submit another PR if you would like to revise and have another review. Please reach out if questions.

Ready to merge

________________________________

In students/CCSt130/lesson03/mail_room_pt2.py<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FUWPCE-PythonCert-ClassRepos%2FSP_Online_PY210%2Fpull%2F339%23discussion_r311856083&data=02%7C01%7C%7C921f6ca520154e5b6ef008d71bbf43a4%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637008380595026513&sdata=kwPIKuZeEiBKKcOXl%2FV0Qyc2JAFWIIdk%2FFMQioEno0k%3D&reserved=0>:

> +# -*- coding: utf-8 -*-
+""" This code allows a user to update a dict of donors, print
+donations, average donations, print a thank you email and letters.
+Per requirements, lists were replaced with dicts where possible.
+"""
+
+"""
+Lesson04 :: Mailroom Part 2
+@author: Chuck Stevens :: CCSt130
+Edited Sun Aug 4 17:31:12 2019
+
+"""
+
+import datetime # dates needed for comparison
+
+from collections import Counter # Count of donations


Make sure imports are at top of file, with no blank lines between

________________________________

In students/CCSt130/lesson03/mail_room_pt2.py<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FUWPCE-PythonCert-ClassRepos%2FSP_Online_PY210%2Fpull%2F339%23discussion_r311856178&data=02%7C01%7C%7C921f6ca520154e5b6ef008d71bbf43a4%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637008380595046549&sdata=uHtQM%2FxmYTpmn1zxFUqakuhNITk8%2BaKBCTfn2NUpX5s%3D&reserved=0>:

> +        if usr_input in quit_opt:
+            quit_menu(exit_msg)
+            break
+        elif usr_input == '4':
+            quit_menu(exit_msg)
+            break
+        # Error handling
+        if not usr_input in menu_dict:
+            print()
+            print("Invalid entry--try again!")
+            print()
+        # Launch a valid option
+        # This will unintentionally print primary keys w/o elif '4'
+        else: menu_dict.get(usr_input)(all_donations)
+
+def quit_menu(exit_strings):


You'll want 2 blank lines between functions

________________________________

In students/CCSt130/lesson03/mail_room_pt2.py<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FUWPCE-PythonCert-ClassRepos%2FSP_Online_PY210%2Fpull%2F339%23discussion_r311856380&data=02%7C01%7C%7C921f6ca520154e5b6ef008d71bbf43a4%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637008380595056550&sdata=fu2RlwFmprUmlNCRS3NxgvaSaAbKrs9V94FzxxJbqIo%3D&reserved=0>:

> +                   \nThank you for your donations to FfFH and your dedication to Feline Health. \
+                   \nYour donations will be put to good use keeping kitties healthy! ",
+                   'Sig_block': "\nBlessings,\n\n\n\nChuck Stevens\nDirector Emeritus\nFoundation for Feline Health"}
+    # Print salutation and body of email
+    print(donor_email['Salutation'].format(a_donor))
+    print(donor_email['Body'])
+
+    # List all relevant donations in the email generated
+    list_donor_donations(all_donations, a_donor)
+
+    # Print signature block
+    print(donor_email['Sig_block'])
+
+    return
+
+def prn_ltrs(all_donations):


Ideally, your function names and variables should be self explanatory and explicit. i.e., print_letters()

________________________________

In students/CCSt130/lesson03/mail_room_pt2.py<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FUWPCE-PythonCert-ClassRepos%2FSP_Online_PY210%2Fpull%2F339%23discussion_r311856447&data=02%7C01%7C%7C921f6ca520154e5b6ef008d71bbf43a4%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637008380595066556&sdata=y%2FE37NJeXZkP831GA7uIUzIzJVub%2Bmtvjh8JlXOJDE0%3D&reserved=0>:

> +    # Print salutation and body of email
+    print(donor_email['Salutation'].format(a_donor))
+    print(donor_email['Body'])
+
+    # List all relevant donations in the email generated
+    list_donor_donations(all_donations, a_donor)
+
+    # Print signature block
+    print(donor_email['Sig_block'])
+
+    return
+
+def prn_ltrs(all_donations):
+    """ Creates formatted letters with donation details. """
+    # For today's date to be printed in email
+    from datetime import date


Go ahead and add this to the imports at the top of the file

________________________________

In students/CCSt130/lesson03/mail_room_pt2.py<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FUWPCE-PythonCert-ClassRepos%2FSP_Online_PY210%2Fpull%2F339%23discussion_r311856514&data=02%7C01%7C%7C921f6ca520154e5b6ef008d71bbf43a4%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637008380595086575&sdata=kYgu4EFxoRIUfa5mUliWDYTHSUH%2Ff3Ebb3zvjoQuV58%3D&reserved=0>:

> +            # Signature block
+            donor_ltr.write(donor_email['Sig_block'])
+            # Not needed
+            # donor_ltr.close() # Close file
+            print()
+            # Informs letter generated
+            print("Donor letter has been saved to current directory for: {}!".format(a_donor))
+
+    return
+
+def prn_dict(a_dict):
+    """ Generic dictionary printer """
+    for k, v in a_dict.items():
+        print(k,":",v)
+
+def findKey(ttl):


This should be snake_case, just like the rest of your functions and variables

________________________________

In students/CCSt130/lesson03/mail_room_pt2.py<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FUWPCE-PythonCert-ClassRepos%2FSP_Online_PY210%2Fpull%2F339%23discussion_r311856628&data=02%7C01%7C%7C921f6ca520154e5b6ef008d71bbf43a4%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637008380595096581&sdata=Rx%2BLk4ziPne9FM3oJ5bUFHcUA4EPDOLH9RBf4XgJh0U%3D&reserved=0>:

> +                break
+            else:
+                print()
+                print("Amount must be greater than zero and less than a billion--try again!")
+                print()
+        else: # Non-numeric entry
+            print()
+            print("Invalid entry--try again!")
+            print()
+    # Inform user
+    print()
+    print(">>> Donation Has Been Entered for {}. <<<".format(a_donor))
+    print()
+    print("###")
+
+    return(a_donor) # Return user's choice


This function breaks when adding new donors and their donations, do you know why?

**Comment by CCSt130** [comment]
Thank you for the pointers; I’ve made a new PR. To answer your question regarding the break—it was breaking because the new donor was not a primary key.

I needed an ‘if’ to check to see if the key was in the data structure already, then use different syntax to add the key and subkey with values, as opposed to the correct syntax to add a subkey and value to an existing primary key.  Yay, fixed!

Happy Weekend,
Chuck

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Pull request merged**